engagement_database/data_models.py

```diff
@@ -104,3 +105,107 @@ def from_dict(cls, d):
 
     def copy(self):
         return Message.from_dict(self.to_dict())
+
+
+class HistoryEntry(object):
+    def __init__(self, update_path, updated_doc, origin, timestamp):
+        """
+        Represents an entry in the database's history, describing an update to one of the documents.
+
+        :param update_path: Full path in Firestore to the document that was updated.
+        :type update_path: str
+        :param updated_doc: Snapshot of the updated document at the time the update was made.
+                            The document object requires a `to_dict()` method so it can be serialized.
+        :type updated_doc: dict | obj with to_dict() method.
+        :param origin: Origin of this update.
+        :type origin: HistoryEntryOrigin
+        :param timestamp: Timestamp this entry was made in Firestore, or None if it hasn't yet been written to Firestore
+        :type timestamp: datetime.datetime | None
+        """
+        self.update_path = update_path
+        self.updated_doc = updated_doc
+        self.origin = origin
+        self.timestamp = timestamp
+
+    def to_dict(self):
+        return {
+            "update_path": self.update_path,
+            "updated_doc": self.updated_doc.to_dict(),
+            "origin": self.origin.to_dict(),
+            "timestamp": self.timestamp
+        }
+
+    @classmethod
+    def from_dict(cls, d, doc_type=None):
+        """
+        :param d: Dictionary to initialise from.
+        :type d: dict
+        :param doc_type: Type to deserialize the updated_doc to e.g. `Message`. If None, returns the updated_doc in its
+                         serialized form.
+        :type doc_type: class with from_dict() method.
+        :return:
+        :rtype: HistoryEntry
+        """
+        return HistoryEntry(
+            update_path=d["update_path"],
+            updated_doc=d["updated_doc"] if doc_type is None else doc_type.from_dict(d["updated_doc"]),
+            origin=HistoryEntryOrigin.from_dict(d["origin"]),
+            timestamp=d["timestamp"]
+        )
+
+
+class HistoryEntryOrigin(object):
+    def __init__(self, name, project, pipeline, details, commit, line=None):
```

Comment:
I think we also want to have an Identifier of the user launching the origin script

Reply:
great suggestion, will add